### PR TITLE
examples for docs

### PR DESCRIPTION
**Pull request by sh-rp**
### Description
This preview provides a scaffold for our examples. Notes:
* This is based on the remove snip sync PR: https://github.com/dlt-hub/dlt/pull/615
* The full code example of every example is in the /code directory of example in the website/docs folder. 
* update_snippets is updated to sync the examples from the website into /docs/examples. It will copy over every file and take the "example" snippet from each file.
* There is a unique component that all examples share, which is the header with some introductory info and stuff that is shared between all examples. This is an approach we can take in other parts of the docs too to reduce duplication.
* The code example itself is just for illustration purposes for now.

### CODE REVIEW

File: docs/website/docs/examples/_examples-header.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import Admonition from "@theme/Admonition";
+import CodeBlock from '@theme/CodeBlock';
+
+<Admonition>
+    The <span>{props.title}</span> Example is part of the <a href="/docs/examples">DLT Code Examples</a>, a list of comprehensive examples to help you solve real world loading problems with dlt.
```
Comment (rudolfix) [inline docs/website/docs/examples/_examples-header.md:5]: dlt not DLT. please!

File: docs/website/docs/examples/_examples-header.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    The source code for this example can be found in our repository at: <a href={"https://github.com/dlt-hub/dlt/tree/devel/docs/examples/" + props.slug}>{"https://github.com/dlt-hub/dlt/tree/devel/docs/examples/" + props.slug}.</a>
+</Admonition>
+
+## TLDR
+<div>{props.intro}</div>
+
+## Setup: Running this example on your machine
+<CodeBlock language="sh">
```
Comment (rudolfix) [inline docs/website/docs/examples/_examples-header.md:13]: any reason to use this and not
```
```sh
```

File: docs/website/docs/examples/incremental_loading/code/run-snippets.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+
+
+def incremental_snippet() -> None:
```
Comment (rudolfix) [inline docs/website/docs/examples/incremental_loading/code/run-snippets.py:3]: that was ok for snippets but is not OK for examples. examples are intended to be visible in full and be runnable with python name.py.
this is IMO convenient for us but not for the user. what I'd do is an example structure that look like this:
```
docs/examples
  |--incremental_loading
         |-.dlt/
         |-incremental_loading.py
         |-README.md
```
where README.md is your index.md (I hope that's possible). then we give people actual github experience

**Review by rudolfix — CHANGES_REQUESTED** [review]
on top of the comments:
should we have docosaurus directly in /docs? o docs/docs? not in docs/website/docs?
we can just delete archives or skip it from being crawled by docosaursu

File: docs/website/docs/examples/incremental_loading/code/run-snippets.py (hunk 3 above)
Comment (sh-rp) [reply docs/website/docs/examples/incremental_loading/code/run-snippets.py:3]: Well I have built a script which exports this examples as runnable projects to /docs/examples and removes all our snipync directives. So we can have nice tests for us and the user can just run it. You can see this in this PR. All the stuff in docs/examples in generated from the stuff in docs/website/docs/examples

File: docs/website/docs/examples/_examples-header.md (hunk 1 above)
Comment (sh-rp) [reply docs/website/docs/examples/_examples-header.md:5]: alright, I have fixed this in a few other places I have found.

File: Makefile [hunk 4]
```diff
@@ -52,7 +52,7 @@ lint:
 	# $(MAKE) lint-security
 
 test-and-lint-snippets:
-	poetry run mypy --config-file mypy.ini docs/website docs/examples
+	poetry run mypy --config-file mypy.ini --namespace-packages --explicit-package-bases docs/website docs/examples
```
[the remark this answers is not in the document]
Comment (sh-rp) [reply Makefile:55]: I have removed this again

**Comment by sh-rp** [comment]
> on top of the comments:
> should we have docosaurus directly in /docs? o docs/docs? not in docs/website/docs?
> we can just delete archives or skip it from being crawled by docosaursu

Yes, I think that would be nice. But we should do this when there are no open documentation branches.

File: docs/website/docs/examples/_examples-header.md (hunk 2 above)
Comment (sh-rp) [reply docs/website/docs/examples/_examples-header.md:13]: Yes, the variable will not work in the markdown block unfortunately..

File: docs/website/docs/examples/incremental_loading/code/run-snippets.py (hunk 3 above)
Comment (rudolfix) [reply docs/website/docs/examples/incremental_loading/code/run-snippets.py:3]: heh that's even better. didn't see that. sorry for stupid comment :)

File: dlt/common/schema/schema.py [hunk 5]
```diff
@@ -392,7 +392,7 @@ def _coerce_non_null_value(self, table_columns: TTableSchemaColumns, table_name:
                 raise CannotCoerceColumnException(table_name, col_name, py_type, table_columns[col_name]["data_type"], v)
             # otherwise we must create variant extension to the table
             # pass final=True so no more auto-variants can be created recursively
-            # TODO: generate callback so DLT user can decide what to do
+            # TODO: generate callback sodltuser can decide what to do
```
Comment (rudolfix) [inline dlt/common/schema/schema.py:395]: somehow surrounding spaces got eaten. and that happened in many places. I hope you can revert the commit and do that again?

**Review by rudolfix — CHANGES_REQUESTED** [review]
Looks like we are good with it so let's write two examples and merge it. I'll take a look at pokemon and maybe add google sheets example?

File: docs/website/docs/examples/transformers/code/.dlt/config.toml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# @@@DLT_SNIPPET_START example
+# @@@DLT_SNIPPET_START toml
+# @@@DLT_SNIPPET_END toml
+
+[runtime]
+log_level="WARNING"
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply docs/website/docs/examples/transformers/code/.dlt/config.toml:6]: look in performance docs and provide configs for extract parallelism

File: docs/website/docs/examples/transformers/code/run-snippets.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    # retrieve pokemon list
+    @dlt.resource(write_disposition="replace")
+    def pokemon_list() -> Iterable[TDataItem]:
+        """
+        Returns an iterator of pokemon
+        Yields:
+            dict: The pokemon list data.
+        """
+        yield from requests.get(POKEMON_URL).json()["results"]
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply docs/website/docs/examples/transformers/code/run-snippets.py:22]: our examples should be practical

File: docs/website/docs/examples/transformers/code/run-snippets.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        return species_data
+
+
+    # build duck db pipeline
+    pipeline = dlt.pipeline(
+        pipeline_name="pokemon", destination="duckdb", dataset_name="pokemon_data"
+    )
+
+    # the pokemon_list resource does not need to be loaded
+    load_info = pipeline.run([pokemon(), species()])
```
Comment (rudolfix) [inline docs/website/docs/examples/transformers/code/run-snippets.py:60]: ok now it is clear why you have double evaluation. please create a dlt.source that returns 3 resources
- the list resource
- pokemons
- species
use pipes to connect resources to transformers

File: docs/website/docs/examples/transformers/index.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+<Header 
+    intro="In this example you will learn how load a list of pokemone from the pokeapi and with the help of dlt transformers
+    automatically query additional data per retrieved pokemon. You will also learn how to harness parallelism with futures."
+    slug="transformer" />
+
+
+## Using transformers with the pokemon api
+
+For this example we will be loading data from the [Poke Api](https://pokeapi.co/). We will load a list of pokemon from the
```
Comment (rudolfix) [inline docs/website/docs/examples/transformers/index.md:17]: Look what @AstrakhantsevaAA did. 3-5 bullet points with highlights
- we create two transformers that are connected to a single list resource with |
- we load in parallel using requests
- we load in parallel using async
- we configure parallelism in config.toml
- we deselect pokemon list by default to not load this in database
provide relevant docs links

File: docs/website/docs/examples/transformers/index.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    pipeline_name="pokemon", destination="duckdb", dataset_name="pokemon_data"
+)
+
+# the pokemon_list resource does not need to be loaded
+load_info = pipeline.run([pokemon(), species()])
+print(load_info)
+```
+<!--@@@DLT_SNIPPET_END ./code/run-snippets.py::example-->
+
+### Example pokemon list data
```
Comment (rudolfix) [inline docs/website/docs/examples/transformers/index.md:86]: I'd not include that

**Review by rudolfix — CHANGES_REQUESTED** [review]
1. instead of adding zendesk credentials, use the same method we have in verified sources (via google secrets)
also because we rely on credentials to run tests we prevent external contributors from using our CI. could we have a pytest skip that will skip the test if it detects we run on github CI and from the fork (easy - all in env variables)

2. python file names are always "run.py". those module names are used to create sections for config and secrets. let's use a meaningful name ie. zendesk, pokemon
3. title the example so it is clear what it does and what is the main highlight:
- "Get Pokemon details in parallel using transformers"
- "Load zendesk tickets incrementally"

Structure is OK, what I'd change
TLDR -> Summary - One sentence what example does
Highlights
3-5 bullet points with the most interesting stuff or building blocks used. ideal bullet point
- one sentence explains what we do in the example + links the docs page (if applicable) + shows relevant code snippet (if the example is long)

IMO we do not need to include the whole snippet (maybe if it is short)

example:

- we configure incremental loading for the resource that takes start and end date passed from the source (link to incremental docs) and allows it to use Airflow schedule (link to the docs)
```
code snippet
```

File: docs/website/docs/examples/transformers/code/run-snippets.py (hunk 8 above)
Comment (sh-rp) [reply docs/website/docs/examples/transformers/code/run-snippets.py:60]: Ok, I have done that, but you should check that it is done the right way. The tests pass.

File: docs/website/docs/examples/transformers/code/run-snippets.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    @dlt.transformer(data_from=pokemon_list)
+    async def pokemon(pokemon: TDataItem):
+        """
+        Returns an iterator of pokemon deatils
+        Yields:
+            dict: The pokemon full data.
+        """
+        # just return the results, if you yield,
+        # generator will be evaluated in main thread
+        return requests.get(pokemon["url"]).json()
```
[the remark this answers is not in the document]
Comment (sh-rp) [reply docs/website/docs/examples/transformers/code/run-snippets.py:34]: I am not sure what you mean with the inner function. In the chess example there is just the defer decorator and the value is returned.

File: docs/website/docs/examples/transformers/code/run-snippets.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        pipeline_name="pokemon", destination="duckdb", dataset_name="pokemon_data"
+    )
+
+    # the pokemon_list resource does not need to be loaded
+    load_info = pipeline.run([pokemon(), species()])
+    print(load_info)
+    # @@@DLT_SNIPPET_END example
+
+    # test assertions
+    row_counts = pipeline.last_trace.last_normalize_info.row_counts
```
[the remark this answers is not in the document]
Comment (sh-rp) [reply docs/website/docs/examples/transformers/code/run-snippets.py:65]: I prefer it this way, because we can add tests inbetween etc. But I can change it if you like.

File: docs/website/docs/examples/transformers/index.md (hunk 10 above)
Comment (sh-rp) [reply docs/website/docs/examples/transformers/index.md:86]: i like this, but I have removed it for now.

File: docs/website/docs/examples/transformers/code/run-snippets.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        # generator will be evaluated in main thread
+        species_data = requests.get(pokemon["species"]["url"]).json()
+        # optionally add pokemon_id to result json, to later be able
+        # to join tables
+        species_data["pokemon_id"] = pokemon["id"]
+        return species_data
+
+
+    # build duck db pipeline
+    pipeline = dlt.pipeline(
```
[the remark this answers is not in the document]
Comment (sh-rp) [reply docs/website/docs/examples/transformers/code/run-snippets.py]: i have added it, we can make it work with what i have built

File: docs/examples/.dlt/secrets.toml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# here is a file with the secrets for all the example pipelines in `examples` folder
```
Comment (sh-rp) [inline docs/examples/.dlt/secrets.toml:1]: @rudolfix should this be here?

**Review by rudolfix — APPROVED** [review]
LGTM! @sh-rp @AstrakhantsevaAA  thanks for your hard work!
